File: dse20503-project/dse20503-react/dse20503-app/src/services/AppointmentService.js
import axios from 'axios';

const API_URL = 'http://localhost:8080/appointments';
axios.defaults.headers.post['Content-Type'] = 'application/json';

const AppointmentService = {
    // Get all appointments
    getAllAppointments: async () => {
        try {
            const response = await axios.get(`${API_URL}/all`);
            return response.data;
        } catch (error) {
            throw new Error(error.response.data.message || 'Error fetching appointments');
        }
    },

    // Get appointment by ID
    getAppointmentById: async (id) => {
        try {
            const response = await axios.get(`${API_URL}/${id}`);
            return response.data;
        } catch (error) {
            throw new Error(error.response.data.message || 'Error fetching appointment');
        }
    },

    // Create a new appointment
    createAppointment: async (appointment) => {
        try {
            const response = await axios.post(`${API_URL}/create`, appointment);
            return response.data;
        } catch (error) {
            throw new Error(error.response.data.message || 'Error creating appointment');
        }
    },

    // Update an appointment
    updateAppointment: async (id, appointmentDetails) => {
        try {
            const response = await axios.put(`${API_URL}/update/${id}`, appointmentDetails);
            return response.data;
        } catch (error) {
            throw new Error(error.response.data.message || 'Error updating appointment');
        }
    },

    // Delete an appointment
    deleteAppointment: async (id) => {
        try {
            const response = await axios.delete(`${API_URL}/delete/${id}`);
            return response.data;
        } catch (error) {
            throw new Error(error.response.data.message || 'Error deleting appointment');
        }
    }
};

export default AppointmentService;

File: dse20503-project/dse20503-spring/src/main/java/com/example/dse20503_project/service/AppointmentService.java
package com.example.dse20503_project.service;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.dse20503_project.entity.Appointment;
import com.example.dse20503_project.repository.AppointmentRepository;

@Service
public class AppointmentService {

    @Autowired
    private AppointmentRepository appointmentRepository;

    // Get all appointments
    public List<Appointment> getAllAppointments() {
        return appointmentRepository.findAll();
    }

    // Get appointment by id
    public Appointment getAppointmentById(Long id) {
        return appointmentRepository.findById(id).orElse(null);
    }

    // Create a new appointment
    public Appointment createAppointment(Appointment appointment) {
        appointment.setCreatedAt(LocalDateTime.now());
        appointment.setUpdatedAt(LocalDateTime.now());
        return appointmentRepository.save(appointment);
    }

    // Update an existing appointment
    public Appointment updateAppointment(Long id, Appointment appointmentDetails) {
        return appointmentRepository.findById(id)
                .map(existingAppointment -> {
                    // Update appointment date if valid
                    if (appointmentDetails.getAppointmentDate() != null &&
                            appointmentDetails.getAppointmentDate().isAfter(LocalDateTime.now())) {
                        existingAppointment.setAppointmentDate(appointmentDetails.getAppointmentDate());
                    } else if (appointmentDetails.getAppointmentDate() != null) {
                        throw new IllegalArgumentException("Appointment date cannot be in the past.");
                    }

                    // Update notes if provided
                    if (appointmentDetails.getNotes() != null) {
                        existingAppointment.setNotes(appointmentDetails.getNotes());
                    }

                    // Update only updatedAt field
                    existingAppointment.setUpdatedAt(LocalDateTime.now());
                    return appointmentRepository.save(existingAppointment);
                })
                .orElseThrow(() -> new RuntimeException("Appointment not found with id " + id));
    }

    // Delete an appointment
    public void deleteAppointment(Long id) {
        appointmentRepository.deleteById(id);
    }
}